src/input/manager.rs

```diff
@@ -692,22 +745,33 @@ impl Manager {
         id: String,
         device: UdevDevice,
     ) -> Result<(), Box<dyn Error>> {
-        // Check all existing composite devices to see if this device is part of
-        // their config
-        'start: for composite_device in self.composite_devices.keys() {
-            let Some(config) = self.used_configs.get(composite_device) else {
+        // Check all CompositeDevice configs to see if this matches a CompositeDevice.
+        // On a match the CompositeDevice will either be created if possible and not existing already
+        // or the device will be added to an existing one if the CompositeDevice is not full.
+        'start: for config in self.load_device_configs().await {
+            let composite_device = (&config).name.clone();
+            let dev_subsystem = (&device).subsystem();
+
+            log::debug!("Checking if composite device {composite_device} wants current device device: {id:?}");
+
+            // Check to see if this configuration matches the system
+            if !config.has_valid_matches(&self.dmi_data, &self.cpu_info) {
+                log::trace!("Configuration does not match system");
                 continue;
-            };
-            log::debug!("Checking if existing composite device {composite_device:?} with config {:?} is missing device: {id:?}", config.name);
+            }
 
             // If the CompositeDevice only allows a single source device, skip its
             // consideration.
             if config.single_source.unwrap_or(false) {
-                log::trace!("{:?} is a single source device. Skipping.", config.name);
+                log::trace!(
+                    "Skipping composite device {composite_device}: is a single source device"
+                );
                 continue;
             }
             if config.maximum_sources.unwrap_or(0) == 1 {
-                log::trace!("{:?} is a single source device. Skipping.", config.name);
+                log::trace!(
+                    "Skipping composite device {composite_device}: is a single source device"
+                );
```

Comment:
With the removal of the second loop, how will new composite devices be created in these circumstances? The expected behavior should be:
- If a device matches an existing composite device, and it hasn't reached its maximum number of source devices, it should be added to the existing composite device.
- If a device matches an existing composite device, but it HAS reached its maximum, a new composite device should be created.
- Lastly, if no existing composite device exists, a new one should be created.

With these changes, I don't see how a new composite device could be ever be created if an existing one reaches its max.

Reply:
For example, you can follow these steps to see this problem manifested:

- Enable management of all supported input devices: `busctl set-property org.shadowblip.InputPlumber /org/shadowblip/InputPlumber/Manager org.shadowblip.InputManager ManageAllDevices b 1`
- Plug in an XBox 360 controller (or other supported external controller). A Composite Device should be created for that controller.
- Plug in a second controller of the same type. A second Composite Device should be created for that second controller.